### PR TITLE
Allow blacklisting printers from USB autosuspend

### PR DESCRIPTION
**Pull request by rimas-kudelis**
This is my first take at fixing #280.

### CODE REVIEW

File: tlp-functions.in [hunk 1]
```diff
@@ -1853,6 +1853,24 @@ set_usb_suspend () { # activate usb autosuspend for all devices except input and
                         fi # phone
 
                         if [ -z "$exc" ]; then
+                            # check for printers
+                            USB_BLACKLIST_PRINTER=${USB_BLACKLIST_PRINTER:-1} # default is exclude
+
+                            if [ "$USB_BLACKLIST_PRINTER" = "1" ]; then
+                                if [ "$dclass" != "00" ]; then
```
Comment (linrunner) [inline tlp-functions.in]: I checked the [USB Printer Class Spec](http://www.usb.org/developers/docs/devclass_docs/usbprint11a021811.pdf). On p.8 it says that bDeviceClass == 00h, so your check will be negative. I suggest you remove it.

File: tlp-functions.in [hunk 2]
```diff
@@ -1853,6 +1853,24 @@ set_usb_suspend () { # activate usb autosuspend for all devices except input and
...
+                                    # check for cdc subdevices
```
Comment (linrunner) [inline tlp-functions.in]: Comment should read "... printer subdevices"

File: tlp-usb-udev.in [hunk 3]
```diff
@@ -190,6 +190,24 @@ if [ -f $usbdev/power/autosuspend ] || [ -f $usbdev/power/autosuspend_delay_ms ]
         fi # phone
 
         if [ -z "$exc" ]; then
+            # check for printers
+            USB_BLACKLIST_PRINTER=${USB_BLACKLIST_PRINTER:-1} # default is exclude
+
+            if [ "$USB_BLACKLIST_PRINTER" = "1" ]; then
+                if [ "$dclass" != "00" ]; then
```
Comment (linrunner) [inline tlp-usb-udev.in]: See above.

**Comment by rimas-kudelis** [comment]
Sorry, I haven't tested this with my printer, and those errors were obviously incompletely adapted copy-paste.

Pushed a fix for this. Could you tell me if there's a way to quickly test the patch in Ubuntu without building the package or installing to `/usr/local`? Would it be enough to just edit `/lib/udev/tlp-usb-udev`?

**Comment by linrunner** [comment]
Editiing the file is ok. Copying **tlp-usb-udev.in** is not, because make install replaces symbols with paths.

Nevertheless I recommend you build a package. It's really simple. Take a look at the [devel docs](http://linrunner.de/en/tlp/docs/tlp-developer-documentation.html#sourcecode).

**Comment by rimas-kudelis** [comment]
Hm, I patched `/lib/udev/tlp-usb-udev` and `/usr/share/tlp-pm/tlp-functions`, but this hasn't helped – I wasn't able to print until I readded my printer back to the blacklist by its id.

Is it possible to trace what tlp does when a device is connected? Logs or something?

**Comment by linrunner** [comment]
Sure, there is a trace mode: http://linrunner.de/en/tlp/docs/tlp-developer-documentation.html#testing

Did you notice my latest remarks about the wrong comparison operator in your code?

**Comment by linrunner** [comment]
Any news?

**Comment by rimas-kudelis** [comment]
Sorry, I was kind of AFK during the past couple weeks. I've pushed another update, and tested that its equivalent seems to work fine for me on my Ubuntu. Here's the relevant debug output line:
```
Rgp 07 15:48:56 dusty tlp[5658]: udev_usb.on_printer_black: Bus 6 Dev 2 ID 04e8:326c /sys/devices/pci0000:00/0000:00:1d.0/usb6/6-2
```

**Pull request merged**